docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,256 @@
+* ExitInfo.Vector = #PF identifies that a #PF caused an asynchronous exit
+* MISC.EXINFO.MADDR = the linear address that page faulted (analogous to CR2)
+* MISC.EXINFO.ERRCD = Page-Fault Error Code - information about the page fault
+
+This table summarizes kernel actions and info in SSA for enclaves in different fault scenarios. All exceptions considered here happen inside enclave causing AEX, so the kernel will have a chance to convert it to the synchronous callback thru vDSO interface if needed. 
+
+| Fault Condition  | Key #PF PFEC Contents  | Kernel/vDSO Action  | EXITINFO/MISC.EXINFO<br>Information | Enclave Action   |
+|---|---|---|---|---|
+| Access a page which has been swapped out | #PF where PFEC.P=0   |  ELD the page from backing store, ERSUME | Vector = #PF<br>MADDR=address<br>ERRCD(PFEC).P=0   | N/A    |
+| Access Page Mapped PROT_NONE<br>(page that the enclave has not mmap'ed)  | #PF where PFEC.P=0   |  invoke user handler | Vector = #PF<br>MADDR=address<br>ERRCD(PFEC).P=0   | call custom handler for on-demand mapping or abort    |
+| Access Page Mapped PROT_W (Page had been mmap'ed by enclave, but not EAUG'ed)  | #PF where PFEC.P=0   | EAUG and map the page then ERESUME   | Vector = #PF<br>MADDR = faulting address<br>ERRCD(PFEC).P=0  |  N/A  |
+| Page Protection mismatch in PTE| #PF where PFEC.W/R or PFEC.I/D will not match PTE  |  invoke user handler  | Vector = #PF<br>MADDR = faulting address<br>ERRCD(PFEC).W/R or I/D does not match protections   |  call custom handler or abort  |
+| Page Protection mismatch in EPCM| #PF where PFEC.SGX=1  | invoke user handler   | Vector = #PF<br>MADDR = faulting address <br>ERRCD(PFEC).SGX=1<br> ERRCD(PFEC).W/R or I/D does not match protections    |  Check if EMODPE in progress and wait if so, <br>or call custom handler or abort  |
+| Access Page with EPCM.Pending  | #PF where PFEC.SGX=1   | invoke user handler   | Vector = #PF<br>MADDR = faulting address<br>ERRCD(PFEC).SGX=1  | Enclave is operating on the page, the handler should wait on the operation to complete. |
+| Access Page with EPCM.Modified | #PF where PFEC.SGX=1   | invoke user handler   | Vector = #PF<br>MADDR = faulting address<br>ERRCD(PFEC).SGX=1  | Enclave is operating on the page, the handler should wait on the operation to complete. |
+| Access Page with type PT_TRIM  | #PF where PFEC.SGX=1  | invoke user handler  | Vector = #PF<br>MADDR = faulting address<br>ERRCD(PFEC).SGX=1  | Memory Access Error - accessing a TRIM'ed page, call custom handler or abort   |
```

Comment:
What about the condition where the enclave has disappeared (due to power state change)?

Reply:
When the enclave is lost, all ENCLS and ENCLU operations on the enclave will fault. And all of them will follow error paths to an eventual fault on  EENTER reported by kernel to user space. And app should cleanup and reload the enclave. Added a couple of cases into the table and a description of how kernel handle error cases in fault handling.